### PR TITLE
add enum check when decoding in cluster object

### PR DESCRIPTION
**Pull request by yunhanw-google**
#### Problem
https://github.com/project-chip/connectedhomeip/issues/19334

#### Change overview
Add enum check support when decoding in cluster object.
For example, spec version 1 allows values "0, 1"
We select "2" as unknown value.
Spec version 2 defines values 2 and 3 to have a meaning.
When version 2 client sends those to version 1 server, the server will treat both 2 and 3 as 2, which it considers unknown.
The unknown value must not ever be sent on the wire, so the version 1 thing will never send it to a version 2 thing.
Devices implementing different versions of the spec need to be able to interact, and the spec defines how

#### Testing
Add negative test to encode one unknown enum, then decode would set it as the first unknown enum with success.
Add yaml test with similar scenario.

### CODE REVIEW

File: src/app/zap-templates/templates/app/cluster-objects-src.zapt [hunk 1]
```diff
@@ -58,6 +58,11 @@ CHIP_ERROR DecodableType::Decode(TLV::TLVReader &reader) {
         {
             {{#zcl_command_arguments}}
             case to_underlying(Fields::k{{asUpperCamelCase label}}):
+#ifdef CHIP_USE_ENUM_CLASS_FOR_IM_ENUM
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/templates/app/cluster-objects-src.zapt:61]: No, this is wrong.  CHIP_USE_ENUM_CLASS_FOR_IM_ENUM is never defined in practice.

What we need to do is to generate different bodies for the Check functions depending on whether it's a strong enum and whether CHIP_USE_ENUM_CLASS_FOR_IM_ENUM is defined....

File: src/app/zap-templates/zcl/data-model/silabs/ami.xml [hunk 2]
```diff
@@ -201,8 +201,8 @@ limitations under the License.
     <item value="0xA2" name="BatteryCoverRemoved"/>
     <item value="0xA3" name="BatteryCoverClosed"/>
     <item value="0xA4" name="ExcessFlow"/>
-    <item value="0xC0" name="CreditOk"/>
-    <item value="0xC1" name="LowCredit"/>
+    <item value="0xB0" name="CreditOk"/>
+    <item value="0xB1" name="LowCredit"/>
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/silabs/ami.xml:205]: Where is this change coming from?

File: src/controller/java/src/chip/devicecontroller/ResubscriptionAttemptCallback.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ *
+ */
+package chip.devicecontroller;
+
+public interface ResubscriptionAttemptCallback {
+  void onResubscriptionAttempt();
```
Comment (bzbarsky-apple) [inline src/controller/java/src/chip/devicecontroller/ResubscriptionAttemptCallback.java]: What does this have to do with this PR?

File: src/lib/core/CHIPError.h [hunk 4]
```diff
@@ -1394,13 +1394,13 @@ using CHIP_ERROR = ::chip::ChipError;
 #define CHIP_ERROR_LISTENER_ALREADY_STOPPED                    CHIP_CORE_ERROR(0x6a)
 
 /**
- *  @def CHIP_ERROR_UNKNOWN_TOPIC
+ *  @def CHIP_ERROR_UNKNOWN_ENUM
  *
  *  @brief
- *    A topic ID was unknown to the recipient.
+ *    A enum was unknown to the recipient.
```
Comment (bzbarsky-apple) [inline src/lib/core/CHIPError.h:1400]:
```suggestion
 *    A enum value was unknown to the recipient.
```

File: src/lib/core/CHIPError.h [hunk 5]
```diff
@@ -1394,13 +1394,13 @@ using CHIP_ERROR = ::chip::ChipError;
...
  *
  */
-#define CHIP_ERROR_UNKNOWN_TOPIC                               CHIP_CORE_ERROR(0x6b)
+#define CHIP_ERROR_UNKNOWN_ENUM                               CHIP_CORE_ERROR(0x6b)
```
Comment (bzbarsky-apple) [inline src/lib/core/CHIPError.h:1403]:
```suggestion
#define CHIP_ERROR_UNKNOWN_ENUM_VALUE                               CHIP_CORE_ERROR(0x6b)
```

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/controller/java/src/chip/devicecontroller/ResubscriptionAttemptCallback.java (hunk 3 above)
Comment (yunhanw-google) [reply src/controller/java/src/chip/devicecontroller/ResubscriptionAttemptCallback.java]: that has been removed. accidental add

File: src/app/zap-templates/zcl/data-model/silabs/ami.xml (hunk 2 above)
Comment (yunhanw-google) [reply src/app/zap-templates/zcl/data-model/silabs/ami.xml:205]: so I see duplicate 0xC0 0XC1
    <item value="0xC0" name="CreditOk"/>
    <item value="0xC1" name="LowCredit"/>
    <item value="0xC0" name="EmergencyCreditInUse"/>
    <item value="0xC1" name="EmergencyCreditExhausted"/>

Then i modify them as b0 and b1

Comment (bzbarsky-apple) [reply src/app/zap-templates/zcl/data-model/silabs/ami.xml:205]: But these are coming from some spec.  What does that spec say?

**Comment by mrjerryjohns** [comment]
So, what I was hoping for, was not an extra 'check' method.

Rather, I was hoping for the following:
1. Insertion of an extra enum value (e.g 'UnknownValue' at the end of each enums' range).
2. Pivot our existing `Decode` signature for enums from:

```
template <typename X, typename std::enable_if_t<std::is_enum<X>::value, int> = 0>
CHIP_ERROR Decode(TLV::TLVReader & reader, X & x)
{
    return reader.Get(x);
}
```

to something like:
```
template <typename X, typename std::enable_if_t<std::is_enum<X>::value, int> = 0>
CHIP_ERROR Decode(TLV::TLVReader & reader, X & x)
{
    return DecodeEnum(reader, x);
}
```

`cluster-objects.h` would contain a `DecodeEnum` method for each generated enum, and would first validate that the value decoded does match one of N enum values, and if not, return an UnknownValue enum back.

**Comment by bzbarsky-apple** [comment]
We can't add UnknownEnum in a forward-compatible way, right?  New values can get defined.

Why would we want to decode to UnknownEnum instead of just failing decode?

File: src/app/zap-templates/zcl/data-model/silabs/ami.xml (hunk 2 above)
Comment (yunhanw-google) [reply src/app/zap-templates/zcl/data-model/silabs/ami.xml:205]: actually i cannot find this ami and enums in spec......

Comment (bzbarsky-apple) [reply src/app/zap-templates/zcl/data-model/silabs/ami.xml:205]: They're not in the _Matter_ spec.  They're in the Zigbee specs...

File: src/app/zap-templates/templates/app/cluster-objects-src.zapt (hunk 1 above)
Comment (yunhanw-google) [reply src/app/zap-templates/templates/app/cluster-objects-src.zapt]: updated, thanks

File: src/lib/core/CHIPError.h (hunk 4 above)
Comment (yunhanw-google) [reply src/lib/core/CHIPError.h]: done, thanks

File: src/lib/core/CHIPError.h (hunk 5 above)
Comment (yunhanw-google) [reply src/lib/core/CHIPError.h]: done, thanks

File: src/app/zap-templates/templates/app/cluster-enums-src.zapt [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+{{#zcl_clusters}}
+namespace {{asUpperCamelCase name}} {
+{{#zcl_enums}}
+
+{{#if (isWeaklyTypedEnum label)}}
+// Need to convert consumers to using the new enum classes, so we
+// don't just have casts all over.
+#ifdef CHIP_USE_ENUM_CLASS_FOR_IM_ENUM
+{{/if}}
+CHIP_ERROR Check{{asType label}}({{asType label}} val)
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/templates/app/cluster-enums-src.zapt:18]: This is fine, but we should be able to output the Check methods for the weakly typed enums even if CHIP_USE_ENUM_CLASS_FOR_IM_ENUM is not defined.  They just need to look a bit different.  That can be a follow-up, though.

File: src/app/zap-templates/templates/app/cluster-enums-src.zapt [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    {{#zcl_enum_items}}
+    case {{parent.label}}::k{{asUpperCamelCase label}}:
+    {{/zcl_enum_items}}
+        return CHIP_NO_ERROR;
+    default:
+        return CHIP_ERROR_UNKNOWN_ENUM_VALUE;
+    }
+}
+{{#if (isWeaklyTypedEnum label)}}
+#endif
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/templates/app/cluster-enums-src.zapt:31]:
```suggestion
#endif // CHIP_USE_ENUM_CLASS_FOR_IM_ENUM
```

File: src/app/zap-templates/templates/app/cluster-objects-src.zapt [hunk 8]
```diff
@@ -58,6 +59,29 @@ CHIP_ERROR DecodableType::Decode(TLV::TLVReader &reader) {
...
             case to_underlying(Fields::k{{asUpperCamelCase label}}):
+                {{#if_is_enum type}}
+                {{#unless isArray}}
+                {{#if (isWeaklyTypedEnum (asUpperCamelCase type))}}
+// Need to convert consumers to using the new enum classes, so we
+// don't just have casts all over.
+#ifdef CHIP_USE_ENUM_CLASS_FOR_IM_ENUM
+                {{/if}}
+                {{#if isNullable}}
+                VerifyOrReturnError(!{{asLowerCamelCase label}}.IsNull(), CHIP_NO_ERROR);
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/templates/app/cluster-objects-src.zapt:70]: But ... this is not going to decode the actualy value, or indeed any values after this one.  Was this codepath tested?

Also, how does it make any sense to examine `{{asLowerCamelCase label}}` here _before_ we have done the decode into it?

File: src/app/zap-templates/templates/app/cluster-objects-src.zapt [hunk 9]
```diff
@@ -58,6 +59,29 @@ CHIP_ERROR DecodableType::Decode(TLV::TLVReader &reader) {
...
+                ReturnErrorOnFailure(Check{{asUpperCamelCase type}}({{asLowerCamelCase label}}.Value()));
+                {{else}}
+                {{#if isOptional}}
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/templates/app/cluster-objects-src.zapt:73]: What if it's both optional and nullable?

File: src/app/zap-templates/templates/app/cluster-objects-src.zapt [hunk 10]
```diff
@@ -58,6 +59,29 @@ CHIP_ERROR DecodableType::Decode(TLV::TLVReader &reader) {
...
+                {{else}}
+                ReturnErrorOnFailure(Check{{asUpperCamelCase type}}({{asLowerCamelCase label}}));
+                {{/if}}
+                {{/if}}
+                {{#if (isWeaklyTypedEnum (asUpperCamelCase type))}}
+#endif
+                {{/if}}
+                {{/unless}}
+                {{/if_is_enum}}
                 ReturnErrorOnFailure(DataModel::Decode(reader, {{asLowerCamelCase label}}));
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/templates/app/cluster-objects-src.zapt:85]: So I think what we should be doing is that the Check functions should be called from _inside `Decode` for enums_.  That will make it work right for enums as command fields, enums as struct fields, lists of enums, enum-typed attributes, nullable enums, optional enums, etc, etc, etc.

File: src/app/zap-templates/templates/app/cluster-objects-src.zapt [hunk 11]
```diff
@@ -80,6 +104,24 @@ CHIP_ERROR TypeInfo::DecodableType::Decode(TLV::TLVReader &reader, const Concret
...
+        {{#if_is_enum type}}
+        {{#unless isArray}}
+        {{#if (isWeaklyTypedEnum (asUpperCamelCase type))}}
+// Need to convert consumers to using the new enum classes, so we
+// don't just have casts all over.
+#ifdef CHIP_USE_ENUM_CLASS_FOR_IM_ENUM
+        {{/if}}
+        {{#if isNullable}}
+        VerifyOrReturnError(!{{asLowerCamelCase label}}.IsNull(), CHIP_NO_ERROR);
+        ReturnErrorOnFailure(Check{{asUpperCamelCase type}}({{asLowerCamelCase label}}.Value()));
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/templates/app/cluster-objects-src.zapt:116]: Again, none of this makes sense.  I ask again: was this tested?

File: src/app/zap-templates/zcl/data-model/chip/general-diagnostics-cluster.xml [hunk 12]
```diff
@@ -87,7 +87,7 @@ limitations under the License.
     <attribute side="server" code="0x01" define="REBOOT_COUNT" type="INT16U" min="0x0000" max="0xFFFF" writable="false" default="0x0000" optional="false">RebootCount</attribute>
     <attribute side="server" code="0x02" define="UPTIME" type="INT64U" min="0x0000000000000000" max="0xFFFFFFFFFFFFFFFF" writable="false" default="0x0000000000000000" optional="true">UpTime</attribute>
     <attribute side="server" code="0x03" define="TOTAL_OPERATIONAL_HOURS" type="INT32U" min="0x00000000" max="0xFFFFFFFF" writable="false" default="0x00000000" optional="true">TotalOperationalHours</attribute>
-    <attribute side="server" code="0x04" define="BOOT_REASONS" type="ENUM8" writable="false" optional="true">BootReasons</attribute>
+    <attribute side="server" code="0x04" define="BOOT_REASONS_TYPE" type="BootReasonType" writable="false" optional="true">BootReasons</attribute>
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/general-diagnostics-cluster.xml:90]: Can we please make the XML changes in a separate PR from the range enforcement, so those can bre reviewed independently?  They're not related.

File: src/app/zap-templates/templates/app/helper.js [hunk 13]
```diff
@@ -714,6 +714,8 @@ function getPythonFieldDefault(type, options)
 function isWeaklyTypedEnum(label)
 {
   return [
+    "Enum8",
+    "Enum16",
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/templates/app/helper.js:718]: I don't think we want to do this, because the goal is for `isWeaklyTypedEnum` to go away altogether.

There are two separate concepts of weak-vs-strong typing for enums:

1. Is the type something that maps to an actual enum type as opposed to an integer)?  This is tested by `if_is_strongly_typed_chip_enum`.  This should be used any time we are looking at types.
2. For an enum definittion (which is by definition maps to an enum type) does it map to an enum type that is an enum class?  This is tested by `isWeaklyTypedEnum`.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/app/zap-templates/templates/app/cluster-objects-src.zapt (hunk 8 above)
Comment (yunhanw-google) [reply src/app/zap-templates/templates/app/cluster-objects-src.zapt:70]: if it is null, there is no need for further decoding?

Comment (yunhanw-google) [reply src/app/zap-templates/templates/app/cluster-objects-src.zapt:70]: yes, you are right, fixing it.

**Comment by mrjerryjohns** [comment]
> We can't add UnknownEnum in a forward-compatible way, right? New values can get defined.

The value of UnknownEnum is meant to be generated that doesn't collide with any other enum value for a particular revision of schema used and linked against for a given application binary.

If new values get defined, that's ok, because UnknownEnum will act as a catch-all for all values that particular binary cannot understand.

When the application decides to pick up a newer schema, at that time, UnknownEnum will get re-generated to select a _new_ value that now, doesn't collide with any of the existing values.

> Why would we want to decode to UnknownEnum instead of just failing decode?

Almost all applications today will actually stop further processing of the decoded data if they encounter errors today, and we don't want to do that here. This becomes even worse if the enum is embedded deep in a struct somewhere, in which case, the application has really no idea where the decode failed. We want to keep the promise of future compatibility by letting `Decode` succeed, but push the onus on applications to figure out what they want to do with enum values they don't understand, by giving them this 'UnknownValue' as a placeholder.

**Comment by bzbarsky-apple** [comment]
> If new values get defined, that's ok, because UnknownEnum will act as a catch-all for all values that particular binary cannot understand (for that particular enum definition).

Ah, that's fair.

> Almost all applications today will actually stop further processing of the decoded data if they encounter errors today, and we don't want to do that here.

That's a good point.  I guess we should aim to do the same thing we do for integer min/max: decode successfully, then enforce the constraint down the line.

I still think this should all be done inside the Decode function, with EnsureValidishValue (name to be figured out) functions implemented for all our enums, so we do it correctly across the board instead of special-casing things in cluster-objects.

**Comment by bzbarsky-apple** [comment]
> That's a good point. I guess we should aim to do the same thing we do for integer min/max: decode successfully, then enforce the constraint down the line.

That said, the issues we've had in door lock were due to the cluster impl assuming constraints on enums would get enforced in the IM layer (which the spec allows; invalid enum would lead to CONSTRAINT_ERROR for the path), so they were ending up with nonsense behavior....  Perhaps what we should have is that after decode of a cluster object suceeds we do a "enforce constraints" pass on it?

File: src/app/zap-templates/zcl/data-model/silabs/ami.xml (hunk 2 above)
Comment (yunhanw-google) [reply src/app/zap-templates/zcl/data-model/silabs/ami.xml:205]: just search zigbee spec, and have not yet found things related to this one. But technically it is not right here. thanks

File: src/app/zap-templates/templates/app/cluster-enums-src.zapt (hunk 7 above)
Comment (yunhanw-google) [reply src/app/zap-templates/templates/app/cluster-enums-src.zapt]: Done, thanks

File: src/app/zap-templates/templates/app/cluster-objects-src.zapt (hunk 9 above)
Comment (yunhanw-google) [reply src/app/zap-templates/templates/app/cluster-objects-src.zapt]: updated, thanks

File: src/app/zap-templates/templates/app/cluster-objects-src.zapt (hunk 10 above)
Comment (yunhanw-google) [reply src/app/zap-templates/templates/app/cluster-objects-src.zapt]: updated, thanks

File: src/app/zap-templates/zcl/data-model/chip/general-diagnostics-cluster.xml (hunk 12 above)
Comment (yunhanw-google) [reply src/app/zap-templates/zcl/data-model/chip/general-diagnostics-cluster.xml]: removed them from this PR

File: src/app/zap-templates/templates/app/cluster-objects-src.zapt (hunk 11 above)
Comment (yunhanw-google) [reply src/app/zap-templates/templates/app/cluster-objects-src.zapt]: updated, thanks

File: src/app/zap-templates/templates/app/helper.js (hunk 13 above)
Comment (yunhanw-google) [reply src/app/zap-templates/templates/app/helper.js]: removed, thanks

File: src/app/zap-templates/templates/app/cluster-enums-check.zapt [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+        return CHIP_ERROR_UNKNOWN_ENUM_VALUE;
+    }
+}
+{{/zcl_enums}}
+
+{{#zcl_bitmaps}}
+static CHIP_ERROR __attribute__((unused)) CheckValidEnumValue({{asUpperCamelCase ../name}}::{{asType label}} val)
+{
+    switch (val) {
+    using EnumType = {{asUpperCamelCase ../name}}::{{asType label}};
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/templates/app/cluster-enums-check.zapt:42]: This is not an enum, right?

File: src/app/zap-templates/templates/app/cluster-enums-check.zapt [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+    {{#zcl_bitmap_items}}
+    case EnumType::k{{asUpperCamelCase label}}:
+    {{/zcl_bitmap_items}}
+        return CHIP_NO_ERROR;
+    default:
+        return CHIP_ERROR_UNKNOWN_ENUM_VALUE;
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/templates/app/cluster-enums-check.zapt:48]: Won't this error out on any bitmap that has 2 bits set?

File: src/app/zap-templates/templates/app/cluster-enums-check.zapt (hunk 14 above)
Comment (yunhanw-google) [reply src/app/zap-templates/templates/app/cluster-enums-check.zapt]: per offline discussion, just removed bitmap part. thanks

File: src/app/zap-templates/templates/app/cluster-enums-check.zapt (hunk 15 above)
Comment (yunhanw-google) [reply src/app/zap-templates/templates/app/cluster-enums-check.zapt]: per offline discussion, just removed bitmap part. thanks

File: src/app/data-model/Decode.h [hunk 16]
```diff
@@ -48,7 +56,9 @@ CHIP_ERROR Decode(TLV::TLVReader & reader, X & x)
 template <typename X, typename std::enable_if_t<std::is_enum<X>::value, int> = 0>
 CHIP_ERROR Decode(TLV::TLVReader & reader, X & x)
 {
-    return reader.Get(x);
+    ReturnErrorOnFailure(reader.Get(x));
+    x = chip::app::Clusters::EnsureKnownEnumValue(x);
```
Comment (bzbarsky-apple) [inline src/app/data-model/Decode.h:60]:
```suggestion
    x = Clusters::EnsureKnownEnumValue(x);
```

File: src/app/tests/suites/TestCluster.yaml [hunk 17]
```diff
@@ -2786,8 +2801,8 @@ tests:
       command: "readAttribute"
       attribute: "nullable_enum_attr"
       response:
-          saveAs: nullableEnumAttr254
-          value: 254
+          saveAs: nullableEnumAttr4
+          value: 4
```
Comment (bzbarsky-apple) [inline src/app/tests/suites/TestCluster.yaml:2805]: Why wuould this be 4?  It should be 3, just like the previous read, since the write did not happen.

File: src/app/zap-templates/templates/app/cluster-enums.zapt [hunk 18]
```diff
@@ -28,7 +28,7 @@ k{{asUpperCamelCase label}} = {{asHex value 2}},
 {{#if (isWeaklyTypedEnum label)}}
 #else // CHIP_USE_ENUM_CLASS_FOR_IM_ENUM
 using {{asType label}} = EmberAf{{asType label}};
-#endif
+#endif // CHIP_USE_ENUM_CLASS_FOR_IM_ENUM
 {{/if}}
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/templates/app/cluster-enums.zapt:32]: Don't we need to output a kInvalid or something for the `first_unused_enum_value` bit?  How are cluster impls supposed to know that the value is invalid otherwise?

File: src/app/zap-templates/zcl/data-model/silabs/ami.xml (hunk 2 above)
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/silabs/ami.xml:205]: Again, these should not be getting changed: they are defined in a spec somewhere.  If they are causing problems and are not part of Matter, they should be removed, as needed.

Comment (yunhanw-google) [reply src/app/zap-templates/zcl/data-model/silabs/ami.xml:205]: https://github.com/project-chip/connectedhomeip/pull/20659

File: src/app/zap-templates/templates/app/cluster-enums.zapt (hunk 18 above)
Comment (yunhanw-google) [reply src/app/zap-templates/templates/app/cluster-enums.zapt:32]: just generated this kvalue, thanks

File: src/app/data-model/Decode.h (hunk 16 above)
Comment (yunhanw-google) [reply src/app/data-model/Decode.h]: done, thanks

File: src/app/zap-templates/templates/app/cluster-enums.zapt [hunk 19]
```diff
@@ -28,8 +28,9 @@ k{{asUpperCamelCase label}} = {{asHex value 2}},
...
+static {{asType label}} __attribute__((unused)) k{{asType label}}FirstUnusedEnumVal = static_cast<{{asType label}}>({{first_unused_enum_value mode="first_unused"}});
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/templates/app/cluster-enums.zapt:33]: Shouldn't this be in the enum class for the cases when we have an enum class?

File: src/platform/bouffalolab/BL602/args.gni [hunk 20]
```diff
@@ -37,7 +37,7 @@ chip_inet_config_enable_dns_resolver = false
 chip_inet_config_enable_tun_endpoint = false
 chip_inet_config_enable_tcp_endpoint = true
 chip_inet_config_enable_udp_endpoint = true
-
+chip_detail_logging = false
```
Comment (bzbarsky-apple) [inline src/platform/bouffalolab/BL602/args.gni:40]: They ran out of space?

File: src/app/zap-templates/templates/app/cluster-enums.zapt (hunk 19 above)
Comment (yunhanw-google) [reply src/app/zap-templates/templates/app/cluster-enums.zapt:33]: then I have to modify more server logic to explicitly process these new enum value, for example
../../src/app/clusters/diagnostic-logs-server/diagnostic-logs-server.cpp:41:21: error: enumeration value 'kFirstUnusedEnumVal' not handled in switch [-Werror,-Wswitch]
            switch (payload.intent)
                    ^~~~~~~~~~~~~~
../../src/app/CommandHandlerInterface.h:223:13: note: in instantiation of function template specialization 'DiagnosticLogsCommandHandler::InvokeCommand(chip::app::CommandHandlerInterface::HandlerContext &)::(anonymous class)::operator()<chip::app::CommandHandlerInterface::HandlerContext, chip::app::Clusters::DiagnosticLogs::Commands::RetrieveLogsRequest::DecodableType>' requested here
            func(handlerContext, requestPayload);

Comment (bzbarsky-apple) [reply src/app/zap-templates/templates/app/cluster-enums.zapt:33]: As discussed on Slack, that is _exactly_ the point of adding this value: to help catch all the places that need to be checking for it.

Also as discussed, to keep scope sane for this PR we can add the value inside the enum class commented out, and then do a separate PR that uncomments it and fixes up all the places that catches.

File: src/platform/bouffalolab/BL602/args.gni (hunk 20 above)
Comment (yunhanw-google) [reply src/platform/bouffalolab/BL602/args.gni:40]: yes, run out of space

**Review by bzbarsky-apple — APPROVED** [review]

**Review by yufengwangca — APPROVED** [review]

File: src/app/zap-templates/templates/app/cluster-enums.zapt (hunk 19 above)
Comment (bzbarsky-apple) [reply src/app/zap-templates/templates/app/cluster-enums.zapt]: Created https://github.com/project-chip/connectedhomeip/pull/20907 to handle the followup.
